### PR TITLE
NMS-13770: Flow processing pipeline

### PR DESCRIPTION
**Pull request by fooker**
Split the flow enrichment out of the elastic implementation. Use a
common pipeline to process, enrich, make and threshold flows before
passing to the repositroies.

This also makes the kafka flow forwarder a freestanding repository
implementation.

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13770

### CODE REVIEW

**Comment by fooker** [comment]
Smoke runs are here: https://app.circleci.com/pipelines/github/OpenNMS/opennms?branch=jira%2FNMS-13770-smoke

**Comment by fooker** [comment]
@j-white just for feedback on the overall design. OIA is still missing.

**Review by christianpape — CHANGES_REQUESTED** [review]
Not much changes needed. Just a bunch of questions.

File: features/flows/elastic/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 1]
```diff
@@ -26,13 +26,6 @@
             <cm:property name="connTimeout" value="5000" /> <!-- 5 second timeout for Elasticsearch operations -->
             <cm:property name="readTimeout" value="30000" /> <!-- 30 second timeout for Elasticsearch socket reads -->
             <cm:property name="retries" value="0" /> <!-- Disable retries by default -->
-            <cm:property name="enableForwarding" value="false" /> <!-- Disable flow forwarding -->
-            <cm:property name="nodeCache.enabled" value="true" /> <!-- Set to false to disable caching -->
-            <cm:property name="nodeCache.maximumSize" value="10000"/> <!-- Set value for unlimited size -->
-            <cm:property name="nodeCache.expireAfterWrite" value="0"/> <!-- in seconds. Set to 0 to never evict elements -->
-            <cm:property name="nodeCache.recordStats" value="true"/> <!-- Set to false to not expose cache statistics via jmx -->
-            <cm:property name="flowThreshold.stepSizeMs" value="300000"/>
-            <cm:property name="flowThreshold.idleTimeoutMs" value="900000"/>
```
Comment (RangerRick) [inline features/flows/elastic/src/main/resources/OSGI-INF/blueprint/blueprint.xml:35]: Can we document (even if it's just in the release notes) why these are removed, and what the alternatives are now, or why they are no longer necessary?

File: features/flows/elastic/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 2]
```diff
@@ -62,12 +55,6 @@
             <cm:property name="alwaysUseRawForQueries" value="true" />
             <cm:property name="timeRangeDurationAggregateThresholdMs" value="120000" /> <!-- 2 minutes -->
             <cm:property name="timeRangeEndpointAggregateThresholdMs" value="604800000" /> <!-- 7 days -->
-
-            <!-- Enrichment settings -->
-            <cm:property name="clockSkewCorrectionThreshold" value="0" />
-
-            <!-- Flow Mangling -->
-            <cm:property name="mangleScriptPath" value="" />
```
Comment (RangerRick) [inline features/flows/elastic/src/main/resources/OSGI-INF/blueprint/blueprint.xml:70]: same for this; if we're losing tunable settings, we should describe how to replace their functionality, or why you shouldn't need to anymore

File: features/flows/elastic/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 1 above)
Comment (fooker) [reply features/flows/elastic/src/main/resources/OSGI-INF/blueprint/blueprint.xml:35]: They moved to another `blueprint.xml` (https://github.com/OpenNMS/opennms/pull/5040/files#diff-229201c368c8b1e5256d8d761b77208627803535b7c6b459309a12704451ad98) but are still available under the same name.

File: features/flows/elastic/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 2 above)
Comment (fooker) [reply features/flows/elastic/src/main/resources/OSGI-INF/blueprint/blueprint.xml:70]: Same as above

File: features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/FlowDocument.java [hunk 3]
```diff
@@ -684,13 +684,6 @@ public Integer getTos() {
     }
 
     public void setTos(final Integer tos) {
-        if (tos != null) {
-            setDscp((tos & 0b11111100) >> 2);
-            setEcn(tos & 0b00000011);
-        } else {
-            setDscp(null);
-            setEcn(null);
-        }
```
[the remark this answers is not in the document]
Comment (fooker) [reply features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/FlowDocument.java:693]: It relays on this implementation here: https://github.com/OpenNMS/opennms/blob/jira/NMS-13770/features/flows/api/src/main/java/org/opennms/netmgt/flows/api/Flow.java#L224

File: features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/ElasticFlowRepository.java [hunk 4]
```diff
@@ -192,79 +112,33 @@ public FlowBulk() {
...
+                                 final TracerRegistry tracerRegistry,
+                                 final IndexSettings indexSettings) {
         this.client = Objects.requireNonNull(jestClient);
         this.indexStrategy = Objects.requireNonNull(indexStrategy);
-        this.documentEnricher = Objects.requireNonNull(documentEnricher);
-        this.sessionUtils = Objects.requireNonNull(sessionUtils);
-        this.nodeDao = Objects.requireNonNull(nodeDao);
-        this.snmpInterfaceDao = Objects.requireNonNull(snmpInterfaceDao);
         this.identity = identity;
         this.tracerRegistry = tracerRegistry;
```
[the remark this answers is not in the document]
Comment (fooker) [reply features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/ElasticFlowRepository.java:124]: I would assume so. Not changed that. And there is some null-handling code around.

File: features/flows/processing/src/main/java/org/opennms/netmgt/flows/processing/impl/FlowThresholdingImpl.java [hunk 5]
```diff
@@ -389,4 +389,71 @@ public void close() {
...
+    public static class ApplicationKey {
+        public final int iface;
+        public final Direction direction;
+        public final String application;
+
+        public ApplicationKey(final int iface,
+                              final Direction direction,
+                              final String application) {
+            this.iface = iface;
+            this.direction = direction;
```
[the remark this answers is not in the document]
Comment (fooker) [reply features/flows/processing/src/main/java/org/opennms/netmgt/flows/processing/impl/FlowThresholdingImpl.java:402]: Just moved the file around.

**Review by RangerRick — DISMISSED** [review]
I will not pretend to understand the flow code in general, and this is a massive patch, but nothing horrible jumps out at me.

I wish there was a better way to break this into smaller more digestible pieces, but I realize sometimes you just can't when you chop things up like this. :D

File: features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/ElasticFlowRepository.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (christianpape) [reply features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/ElasticFlowRepository.java:124]: ok

File: features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/FlowDocument.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (christianpape) [reply features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/FlowDocument.java:693]: Ah ok. Got it.

File: features/flows/processing/src/main/java/org/opennms/netmgt/flows/processing/impl/FlowThresholdingImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (christianpape) [reply features/flows/processing/src/main/java/org/opennms/netmgt/flows/processing/impl/FlowThresholdingImpl.java:402]: Ok

**Comment by fooker** [comment]
> I will not pretend to understand the flow code in general, and this is a massive patch, but nothing horrible jumps out at me.
> 
> I wish there was a better way to break this into smaller more digestible pieces, but I realize sometimes you just can't when you chop things up like this. :D

I would like this, too. And I tried to do so. First commit is somewhat freestanding but the majority is not.

File: features/flows/processing/src/main/java/org/opennms/netmgt/flows/processing/enrichment/EnrichedFlow.java [hunk 6]
```diff
@@ -0,0 +1,524 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2020 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2020 The OpenNMS Group, Inc.
```
[the remark this answers is not in the document]
Comment (fooker) [reply features/flows/processing/src/main/java/org/opennms/netmgt/flows/processing/enrichment/EnrichedFlow.java]: changed

**Review by christianpape — APPROVED** [review]
LGTM.

**Comment by fooker** [comment]
@RangerRick @christianpape needs review again because of merge conflicts

**Comment by fooker** [comment]
Work for OIA is [here](https://github.com/OpenNMS/opennms/pull/5073) and [here](https://github.com/OpenNMS/opennms-integration-api/pull/63).

**Comment by fooker** [comment]
> I presume this won't get merged until the OIA stuff is merged/approved. But otherwise it looks like your recent modifications aren't massive, so re-approving.

It is already useful on it's own. Therefore I'd like to merge before the OIA part is done.

**Pull request merged**